src/Storages/RabbitMQ/ReadBufferFromRabbitMQConsumer.cpp

```diff
@@ -0,0 +1,469 @@
+    }
+}
+
+
+void ReadBufferFromRabbitMQConsumer::subscribe(const String & queue_name)
+{
+    if (subscribed_queue[queue_name])
+        return;
+
+    consumer_channel->consume(queue_name, AMQP::noack)
+    .onSuccess([&](const std::string & /* consumer */)
+    {
+        subscribed_queue[queue_name] = true;
+        consumer_error = false;
+        ++count_subscribed;
+
+        LOG_TRACE(log, "Consumer {} is subscribed to queue {}", channel_id, queue_name);
+    })
+    .onReceived([&](const AMQP::Message & message, uint64_t /* deliveryTag */, bool /* redelivered */)
+    {
+        size_t message_size = message.bodySize();
+        if (message_size && message.body() != nullptr)
+        {
+            String message_received = std::string(message.body(), message.body() + message_size);
+
+            if (row_delimiter != '\0')
+            {
+                message_received += row_delimiter;
+            }
+
+            /// Needed to avoid data race because this vector can be used at the same time by another thread in nextImpl().
+            {
+                std::lock_guard lock(mutex);
+                received.push_back(message_received);
+            }
+
+            /* As event loop is blocking to the thread that started it and a single thread should not be blocked while
+             * executing all callbacks on the connection (not only its own), then there should be some point to unblock.
+             * loop_started == 1 if current consumer is started the loop and not another.
+             */
+            if (!loop_started.load() && !event_handler.checkStopIsScheduled())
+            {
+                stopEventLoopWithTimeout();
+            }
+        }
+    })
+    .onError([&](const char * message)
+    {
+        consumer_error = true;
+        LOG_ERROR(log, "Consumer {} failed. Reason: {}", channel_id, message);
+    });
+}
+
+
+void ReadBufferFromRabbitMQConsumer::checkSubscription()
+{
+    if (count_subscribed == num_queues)
+        return;
+
+    wait_subscribed = num_queues;
+
+    /// These variables are updated in a separate thread.
+    while (count_subscribed != wait_subscribed && !consumer_error)
+    {
+        startEventLoop(loop_started);
+    }
+
+    LOG_TRACE(log, "Consumer {} is subscribed to {} queues", channel_id, count_subscribed);
+
+    /// Updated in callbacks which are run by the loop.
+    if (count_subscribed == num_queues)
+        return;
+
+    /// A case that should never normally happen.
+    for (auto & queue : queues)
+    {
+        subscribe(queue);
+    }
+}
+
+
+void ReadBufferFromRabbitMQConsumer::stopEventLoop()
+{
+    event_handler.stop();
+}
+
+
+void ReadBufferFromRabbitMQConsumer::stopEventLoopWithTimeout()
+{
+    event_handler.stopWithTimeout();
+}
+
+
+void ReadBufferFromRabbitMQConsumer::startEventLoop(std::atomic<bool> & loop_started)
+{
+    event_handler.startConsumerLoop(loop_started);
+}
+
+
+bool ReadBufferFromRabbitMQConsumer::nextImpl()
```

Comment:
This logic is very complicated and error-prone. We discussed in chat, that it's better to create a separate task to the which will run non-blocking loop iterations and store records in `ConcurrentBoundedQueue` until it's full. Reader threads will just pop messages from the queue and store them in a local buffer. The code should become much simpler.